### PR TITLE
Introduce Bookmark Manager

### PR DESCRIPTION
**Pull request by bigmontz**
Neo4j clusters are causally consistent, meaning that by default there is no guarantee a query will be able to read changes made by a previous query. For cases where such a guarantee is necessary, the server provides bookmarks to the client. A bookmark is an abstract token that represents some state of the database. By passing one or multiple bookmarks along with a query, the server will make sure that the query will not get executed before the represented state(s) (or a later state) have been established.

The bookmark manager is an interface used by the driver for keeping track of the bookmarks and this way keeping sessions automatically consistent.

### Usage in the session

Enabling it is done by supplying an BookmarkManager implementation instance to this param.
A default implementation could be acquired by calling the factory function `neo4j.bookmarkManager()`.

**Warning**: Share the same BookmarkManager instance across all session can have a negative impact
on performance since all the queries will wait for the latest changes being propagated across the cluster.
For keeping consistency between a group of queries, use `Session` for grouping them.
For keeping consistency between a group of sessions, use `BookmarkManager` instance for grouping them.

```javascript
const bookmarkManager = neo4j.bookmarkManager()
const linkedSession1 = driver.session({ database:'neo4j', bookmarkManager })
const linkedSession2 = driver.session({ database:'neo4j', bookmarkManager })
const unlinkedSession = driver.session({ database:'neo4j' })

// Creating Driver User
const createUserQueryResult = await linkedSession1.run('CREATE (p:Person {name: $name})', { name: 'Driver User'})

// Reading Driver User will *NOT* wait of the changes being propagated to the server before RUN the query
// So the 'Driver User' person might not exist in the Result
const unlinkedReadResult = await unlinkedSession.run('CREATE (p:Person {name: $name}) RETURN p', { name: 'Driver User'})

// Reading Driver User will wait of the changes being propagated to the server before RUN the query
// So the 'Driver User' person should exist in the Result, unless deleted.
const linkedSession2 = await linkedSession2.run('CREATE (p:Person {name: $name}) RETURN p', { name: 'Driver User'})

await linkedSession1.close()
await linkedSession2.close()
await unlinkedSession.close()
```

### CODE REVIEW

File: packages/core/src/bookmark-manager.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+ */
+export default class BookmarkManager {
+  constructor () {
+    throw new Error('Not implemented')
+  }
+
+  /**
+   * Method called when the bookmarks get updated when a transaction finished.
+   *
+   * This method will be called during when auto-commit queries finished and explicit transactions
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
   * This method will be called when auto-commit queries finish and when explicit transactions
```

File: packages/core/src/bookmark-manager.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+   * get commited.
+   * @param {string} database The database which the bookmarks belongs to
+   * @param {string[]} previousBookmarks The bookmarks used during the transaction creation
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
   * @param {string[]} previousBookmarks The bookmarks used when starting the transaction
```

File: packages/core/src/bookmark-manager.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+   * @param {string[]} newBookmarks The new bookmarks resolved at the end of the transaction.
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
   * @param {string[]} newBookmarks The new bookmarks received at the end of the transaction.
```

File: packages/core/src/bookmark-manager.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+   * @returns {void}
+  */
+  updateBookmarks (database: string, previousBookmarks: string[], newBookmarks: string[]): void {
+    throw new Error('Not implemented')
+  }
+
+  /**
+   * Method called by the driver to get the bookmark for one specific database
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
   * Method called by the driver to get the bookmarks for one specific database
```

File: packages/core/src/bookmark-manager.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+   *
+   * @param {string} database The database which the bookmarks belongs to
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
   * @param {string} database The database which the bookmarks belong to
```

File: packages/core/src/bookmark-manager.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+   * @returns {string[]} The set of bookmarks
+   */
+  getBookmarks (database: string): string[] {
+    throw new Error('Not implemented')
+  }
+
+  /**
+   * Method called by the driver for getting all the bookmarks.
+   *
+   * The return of this method should be all the bookmarks present in the BookmarkManager for all databases.
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
   * This method should return all bookmarks for all databases present in the BookmarkManager.
```

File: packages/core/src/bookmark-manager.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    newBookmarks.forEach(bm => bookmarks.add(bm))
+
+    if (typeof this._bookmarksConsumer === 'function') {
+      this._bookmarksConsumer(database, [...bookmarks])
+    }
+  }
+
+  private _getOrInitializeBookmarks (database: string): Set<string> {
+    let maybeBookmarks = this._bookmarksPerDb.get(database)
+    if (maybeBookmarks == null) {
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]: Why `maybeBookmarks == null` over `maybeBookmarks === undefined`?

File: packages/core/src/bookmark-manager.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    }
+    return maybeBookmarks
+  }
+
+  getBookmarks (database: string): string[] {
+    const bookmarks = this._bookmarksPerDb.get(database) ?? []
+
+    if (typeof this._bookmarksSupplier === 'function') {
+      const suppliedBookmarks = this._bookmarksSupplier(database) ?? []
+      return [...bookmarks, ...suppliedBookmarks]
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]: Should we use a `Set` to get rid of potential duplicates that the bookmark supplier might give us?

File: packages/core/src/bookmark-manager.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    return [...bookmarks]
+  }
+
+  getAllBookmarks (): string[] {
+    const bookmarks = []
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]: Should we use a `Set` here too to also get rid of potential duplicates that the bookmark supplier might give us?

File: packages/core/src/driver.ts [hunk 10]
```diff
@@ -87,13 +88,15 @@ type CreateSession = (args: {
   reactive: boolean
   fetchSize: number
   impersonatedUser?: string
+  bookmarkManager?: BookmarkManager
```
Comment (robsdedude) [inline packages/core/src/driver.ts:91]: That's not what the ADR specifies, is it? :thinking:  
I would've expected an `ignoreBookmarkManager` option instead.

Which is also what I see in the implementation below.

File: packages/core/src/index.ts [hunk 11]
```diff
@@ -214,7 +217,9 @@ export type {
   NotificationPosition,
   QueryResult,
   ResultObserver,
-  TransactionConfig
+  TransactionConfig,
+  BookmarkManager,
+  BookmarkManagerConfig
```
Comment (robsdedude) [inline packages/core/src/index.ts]: Probably too late to mention it here as this is how it's spec'ed out in the ADR, but it feels a slightly off that `BookmarkManagerConfig` is not the config for `BookmarkManager`, but only for the default implementation of the `BookmarkManager` provided by the driver lib.

File: packages/core/src/session.ts [hunk 12]
```diff
@@ -332,6 +338,11 @@ class Session {
     return this._lastBookmarks.values()
   }
 
+  private _bookmarks (): Bookmarks {
+    const bookmarks = this._bookmarkManager?.getAllBookmarks() ?? []
+    return new Bookmarks([...bookmarks, ...this._lastBookmarks])
```
Comment (robsdedude) [inline packages/core/src/session.ts]: If I read this all correctly, this is not exactly how the ADR describes it:
Once the session received bookmarks from the server, it should only rely on the bookmarks of the BMM, but no longer on its own bookmarks. This will make a difference, if there is another session (`s2`) running for the same db that fetches `s1._lastBookmarks` via the BMM and completes a TX replacing `s1._lastBookmarks` with `s2._lastBookmarks` in the BMM. This code will still send `s1._lastBookmarks` over the wire. It's not incorrect as the server will ignore the outdated bookmarks but it's wasteful still.

File: packages/core/src/session.ts [hunk 13]
```diff
@@ -283,14 +289,14 @@ class Session {
...
-      onBookmarks: this._updateBookmarks.bind(this),
+      onBookmarks: (newBm, oldBm, db) => this._updateBookmarks(newBm, oldBm, db),
       onConnection: this._assertSessionIsOpen.bind(this),
       reactive: this._reactive,
       fetchSize: this._fetchSize,
       lowRecordWatermark: this._lowRecordWatermark,
       highRecordWatermark: this._highRecordWatermark
     })
-    tx._begin(this._lastBookmarks, txConfig)
+    tx._begin(() => this._bookmarks(), txConfig)
```
Comment (robsdedude) [inline packages/core/src/session.ts:302]: Why do we deffer the bookmark fetching?

File: packages/core/test/bookmark-manager.test.ts [hunk 14]
```diff
@@ -0,0 +1,213 @@
+/**
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts:1]: TODO @robsdedude: continue review from here

File: packages/core/test/bookmark-manager.test.ts [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+        manager.getAllBookmarks(),
+        newBookmarks
+      )
+
+      expect(bookmarksConsumer).toBeCalledWith('neo4j', newBookmarks)
+    })
+  })
+
+  describe('forget()', () => {
+    it('should forgot database', () => {
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts]:
```suggestion
    it('should forget database', () => {
```

File: packages/core/test/bookmark-manager.test.ts [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        'neo4j',
+        manager.getAllBookmarks(),
+        newBookmarks
+      )
+
+      expect(bookmarksConsumer).toBeCalledWith('neo4j', newBookmarks)
+    })
+  })
+
+  describe('forget()', () => {
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts:253]: Consider adding a test for `forget("nonexistent-db")` to see if the BMM can handle that.

File: packages/core/test/bookmark-manager.test.ts [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  bookmarkManager
+} from '../src/bookmark-manager'
+
+describe('BookmarkManager', () => {
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts:24]: For Python I also added tests to check if duplicate BMs are being removed (duplicates in the initial bookmarks, duplicates from the supplier, duplicates that are both in the BMM and being supplied by the supplier). I'm not sure this even applies to the JS BMM, but if it does, it's be worth adding tests for it.

File: packages/core/test/bookmark-manager.test.ts [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+          ['system', systemBookmarks]
+        ])
+      })
+
+      const bookmarks = manager.getAllBookmarks()
+
+      expect(bookmarks).toEqual([...neo4jBookmarks, ...systemBookmarks])
+    })
+
+    it('should return empty if there isnt bookmarks for any db', () => {
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts]:
```suggestion
    it('should return empty if there are no bookmarks for any db', () => {
```

File: packages/core/test/bookmark-manager.test.ts [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+      const manager = bookmarkManager({})
+
+      const bookmarks = manager.getAllBookmarks()
+
+      expect(bookmarks).toEqual([])
+    })
+
+    it('should return enrich bookmarks list with supplied bookmarks', () => {
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts]:
```suggestion
    it('should return enriched bookmarks list with supplied bookmarks', () => {
```

File: packages/core/src/bookmark-manager.ts [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+  bookmarksSupplier?: (database?: string) => string[]
+  bookmarksConsumer?: (database: string, bookmarks: string[]) => void
+}
+
+/**
+ * @typedef {Object} BookmarkManagerConfig
+ * @property {Map<string,string[]>} [initialBookmarks] Defines the initial set of bookmarks. The key is the database name and the values are the bookmarks.
+ * @property {function([database]: string):string[]} [bookmarksSupplier] Called for supplying extra bookmarks to the BookmarkManager
+ * 1. supplying bookmarks from the given database when the default BookmarkManager's `.getBookmarks(database)` gets called.
+ * 2. supplying all the bookmarks when the default BookmarkManager's  `.getAllBookmarks()` gets called
```
Comment (robsdedude) [inline packages/core/src/bookmark-manager.ts]:
```suggestion
 * 2. supplying all the bookmarks when the default BookmarkManager's `.getAllBookmarks()` gets called
```

File: packages/core/test/bookmark-manager.test.ts [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+        initialBookmarks: new Map([
+          ['neo4j', neo4jBookmarks],
+          ['system', systemBookmarks]
+        ]),
+        bookmarksSupplier
+      })
+
+      const bookmarks = manager.getAllBookmarks()
+
+      expect(bookmarks.sort()).toEqual(
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts]: Might be premature optimization but you could do the `.sort()`ed comparison for all tests because we don't really care for the order of the bookmarks. So should that implementation detail ever change, will the future maintainer not wonder if the order is relevant or not. Plus, they won't have to change the test code.

File: packages/core/test/bookmark-manager.test.ts [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+      })
+
+      manager.getAllBookmarks()
+
+      expect(bookmarksSupplier).toBeCalledWith()
+    })
+  })
+
+  describe('updateBookmarks()', () => {
+    it('should remove previous bookmarks and new bookmarks an existing db', () => {
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts]:
```suggestion
    it('should remove previous bookmarks and new bookmarks for an existing db', () => {
```

File: packages/core/test/bookmark-manager.test.ts [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+      const manager = bookmarkManager({
+        initialBookmarks: new Map([
+          ['neo4j', neo4jBookmarks],
+          ['system', systemBookmarks]
+        ])
+      })
+
+      manager.updateBookmarks(
+        'neo4j',
+        manager.getAllBookmarks(),
```
Comment (robsdedude) [inline packages/core/test/bookmark-manager.test.ts]: Could be worth adding a previous bookmark that does not exist in the BMM. In the real world this could also happen if another tx has caused the BMM to remove a BM that was still there when the first tx got started.

File: packages/core/test/session.test.ts [hunk 24]
```diff
@@ -310,6 +314,203 @@ describe('session', () => {
...
+          bookmarks: new bookmarks.Bookmarks([...neo4jBookmarks, ...systemBookmarks, ...lastBookmarks])
+        })
+      ])
+    })
+
+    it.each([
+      [undefined],
+      ['neo4j'],
+      ['adb']
+    ])('should not call run updateBookmarks when bookmarks since it didnt have bookmarks', async (metaDb) => {
```
Comment (robsdedude) [inline packages/core/test/session.test.ts]:
```suggestion
    ])('should not call updateBookmarks when server returns no bookmarks', async (metaDb) => {
```

File: packages/core/test/session.test.ts [hunk 25]
```diff
@@ -310,6 +314,203 @@ describe('session', () => {
...
+      expect(updateBookmarksSpy).not.toBeCalled()
+    })
+  })
+
+  describe('.commit()', () => {
+    it.each([
+      [undefined, 'neo4j'],
+      ['neo4j', 'neo4j'],
+      ['adb', 'adb']
+    ])('should call run updateBookmarks when bookmarks is not empty', async (metaDb, updateDb) => {
```
Comment (robsdedude) [inline packages/core/test/session.test.ts]:
```suggestion
    ])('should call updateBookmarks when server returns non-empty bookmarks', async (metaDb, updateDb) => {
```

File: packages/core/test/session.test.ts [hunk 26]
```diff
@@ -310,6 +314,203 @@ describe('session', () => {
...
+
+      expect(updateBookmarksSpy).toBeCalledTimes(1)
+      expect(updateBookmarksSpy).toBeCalledWith(updateDb, [...neo4jBookmarks, ...systemBookmarks], customBookmarks)
+    })
+
+    it.each([
+      [undefined],
+      ['neo4j'],
+      ['adb']
+    ])('should not call run updateBookmarks when bookmarks is empty', async (metaDb) => {
```
Comment (robsdedude) [inline packages/core/test/session.test.ts]:
```suggestion
    ])('should not call updateBookmarks when server returns no bookmarks', async (metaDb) => {
```

File: packages/core/test/session.test.ts [hunk 27]
```diff
@@ -310,6 +314,203 @@ describe('session', () => {
...
+      const connection = mockBeginWithSuccess(newFakeConnection())
+
+      const { session } = setupSession({
+        connection,
+        bookmarkManager: manager,
+        beginTx: false,
+        database: 'neo4j'
+      })
+
+      await session.run('query')
```
Comment (robsdedude) [inline packages/core/test/session.test.ts]: Were in `describe('.beginTransaction()'`. What is this test doing here?

File: packages/core/test/session.test.ts [hunk 28]
```diff
@@ -349,45 +550,254 @@ describe('session', () => {
...
+          bookmarks: new bookmarks.Bookmarks([...neo4jBookmarks, ...systemBookmarks, ...lastBookmarks])
+        })
+      )
+    })
+
+    it.each([
+      [undefined, 'neo4j'],
+      ['neo4j', 'neo4j'],
+      ['adb', 'adb']
+    ])('should call run updateBookmarks when bookmarks is not empty', async (metaDb, updateDb) => {
```
Comment (robsdedude) [inline packages/core/test/session.test.ts]:
```suggestion
    ])('should call updateBookmarks when server returns non-empty bookmarks', async (metaDb, updateDb) => {
```

File: packages/core/test/session.test.ts [hunk 29]
```diff
@@ -349,45 +550,254 @@ describe('session', () => {
...
+      afterComplete({ db: metaDb, bookmark: customBookmarks })
+
+      expect(updateBookmarksSpy).toBeCalledWith(updateDb, [...neo4jBookmarks, ...systemBookmarks], customBookmarks)
+    })
+
+    it.each([
+      [undefined],
+      ['neo4j'],
+      ['adb']
+    ])('should not call run updateBookmarks when bookmarks is empty', async (metaDb) => {
```
Comment (robsdedude) [inline packages/core/test/session.test.ts]:
```suggestion
    ])('should not call updateBookmarks when server returns no bookmarks', async (metaDb) => {
```

File: packages/neo4j-driver-lite/src/index.ts [hunk 30]
```diff
@@ -199,6 +202,21 @@ const {
  *         logger: (level, message) => console.log(level + ' ' + message)
  *       },
  *
+ *       // Configure a BookmarkManager for in the driver
```
Comment (robsdedude) [inline packages/neo4j-driver-lite/src/index.ts]:
```suggestion
 *       // Configure a BookmarkManager for the driver to use
```

File: packages/neo4j-driver-lite/src/index.ts [hunk 31]
```diff
@@ -199,6 +202,21 @@ const {
...
+ *       //
+ *       // BookmarkManager is a piece of software responsible for keeping casual concistency between different sessions by sharing bookmarks
```
Comment (robsdedude) [inline packages/neo4j-driver-lite/src/index.ts]:
```suggestion
 *       // A BookmarkManager is a piece of software responsible for keeping casual consistency between different sessions by sharing bookmarks
```

File: packages/neo4j-driver-lite/src/index.ts [hunk 32]
```diff
@@ -199,6 +202,21 @@ const {
...
+ *       // between the them.
+ *       // Enabling it is done by supplying an BookmarkManager implementation instance to this param.
+ *       // A default implementation could be acquired by calling bookmarkManager factory function.
```
Comment (robsdedude) [inline packages/neo4j-driver-lite/src/index.ts]:
```suggestion
 *       // A default implementation could be acquired by calling the factory function bookmarkManager.
```

File: packages/neo4j-driver-lite/src/index.ts [hunk 33]
```diff
@@ -199,6 +202,21 @@ const {
...
+ *       //
+ *       // **Warning**: Enabling Bookmark Manager can have a negative impact performance wise since all the queries will wait for the last changes
```
Comment (robsdedude) [inline packages/neo4j-driver-lite/src/index.ts]:
```suggestion
 *       // **Warning**: Enabling the BookmarkManager can have a negative impact on performance since all the queries will wait for the latest changes
```

File: packages/neo4j-driver-lite/src/index.ts [hunk 34]
```diff
@@ -199,6 +202,21 @@ const {
...
+ *       // being propagated accross the cluster.
+ *       // For keeping concistency between a group of queries, use Session for grouping them.
```
Comment (robsdedude) [inline packages/neo4j-driver-lite/src/index.ts]:
```suggestion
 *       // For keeping consistency between a group of queries, use Session for grouping them.
```

File: packages/neo4j-driver/src/index.js [hunk 35]
```diff
@@ -185,6 +186,21 @@ const {
  *         logger: (level, message) => console.log(level + ' ' + message)
  *       },
  *
+ *       // Configure a BookmarkManager for in the driver
```
Comment (robsdedude) [inline packages/neo4j-driver/src/index.js]: See comments on the same documentation in the lite driver

File: packages/neo4j-driver-lite/src/index.ts [hunk 36]
```diff
@@ -199,6 +202,21 @@ const {
...
  *
+ *       // Configure a BookmarkManager for in the driver
+ *       //
+ *       // BookmarkManager is a piece of software responsible for keeping casual concistency between different sessions by sharing bookmarks
+ *       // between the them.
+ *       // Enabling it is done by supplying an BookmarkManager implementation instance to this param.
+ *       // A default implementation could be acquired by calling bookmarkManager factory function.
+ *       //
+ *       // **Warning**: Enabling Bookmark Manager can have a negative impact performance wise since all the queries will wait for the last changes
+ *       // being propagated accross the cluster.
```
Comment (robsdedude) [inline packages/neo4j-driver-lite/src/index.ts]:
```suggestion
 *       // being propagated across the cluster.
```

File: packages/core/src/bookmark-manager.ts (hunk 7 above)
Comment (bigmontz) [reply packages/core/src/bookmark-manager.ts]: `== null` includes `undefined` (actually, means `null` or `undefined`), but I will change for `=== undefined` since the method `Map.prototype.get()` returns `undefined` for non-existing keys.

File: packages/core/src/bookmark-manager.ts (hunk 9 above)
Comment (bigmontz) [reply packages/core/src/bookmark-manager.ts]: Yes, I was thinking about it during the implementation. The methods of `Set` are not that ergonomic as the ones from `Array` in JS and the value will transformed to `Bookmarks`, which removes duplicates. However, for an external user, it could a bit of pain have duplicates in this list.

File: packages/core/src/driver.ts (hunk 10 above)
Comment (bigmontz) [reply packages/core/src/driver.ts:91]: This is the internal api used by the `driver.session()`, see `driver.session()` and `driver._newSession`

File: packages/core/src/index.ts (hunk 11 above)
Comment (bigmontz) [reply packages/core/src/index.ts]: Yes, it is the config for the `neo4j.bookmarkManager` factory. Different configs could also creates instances of different classes.

File: packages/core/src/session.ts (hunk 13 above)
Comment (bigmontz) [reply packages/core/src/session.ts:302]: Because we are only sure about the bookmarks after acquire the connection, it is a bit of leftover from the version where we needed to `manager.getAllBookmarks(['system', session.database])`

File: packages/core/src/session.ts (hunk 12 above)
Comment (bigmontz) [reply packages/core/src/session.ts]: Sure, the ADR was changed after the code is already implemented.

File: packages/core/test/session.test.ts (hunk 27 above)
Comment (bigmontz) [reply packages/core/test/session.test.ts]: I will check it out.

File: packages/core/src/bookmark-manager.ts (hunk 8 above)
Comment (bigmontz) [reply packages/core/src/bookmark-manager.ts]: Implementation has been updated.

File: packages/core/src/session.ts (hunk 13 above)
Comment (bigmontz) [reply packages/core/src/session.ts:302]: Actually, now it is needed, because the begin method is not async

File: packages/core/src/driver.ts (hunk 10 above)
Comment (robsdedude) [reply packages/core/src/driver.ts:91]: Ah, thanks!

File: packages/core/test/utils/matchers.ts [hunk 37]
```diff
@@ -0,0 +1,43 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    interface Matchers<R> {
+      toBeSortedEqual: (expected: Iterable<string>) => CustomMatcherResult
+    }
+  }
+}
+
+export function installMatchers (): void {
+  expect.extend({
+    toBeSortedEqual (this: jest.MatcherContext, received: Iterable<string>, expected: Iterable<string>): jest.CustomMatcherResult {
```
Comment (robsdedude) [inline packages/core/test/utils/matchers.ts:32]: could this be
```suggestion
    toBeSortedEqual (this: jest.MatcherContext, received: Iterable<any>, expected: Iterable<any>): jest.CustomMatcherResult {
```

to be of more general use for the *future*™?

**Review by robsdedude — APPROVED** [review]
:bookmark: :bowtie:

File: packages/core/test/utils/matchers.ts (hunk 37 above)
Comment (bigmontz) [reply packages/core/test/utils/matchers.ts:32]: It was causing type issues™ and I'd like to let myself of the future solve it :D

File: packages/core/src/driver.ts [hunk 38]
```diff
@@ -96,7 +96,110 @@ interface DriverConfig {
...
+ * @param {string} defaultAccessMode=WRITE - The access mode of this session, allowed values are {@link READ} and {@link WRITE}.
+ * @param {string|string[]} bookmarks - The initial reference or references to some previous
+ * transactions. Value is optional and absence indicates that that the bookmarks do not exist or are unknown.
+ * @param {number} fetchSize - The record fetch size of each batch of this session.
+ * Use {@link FETCH_ALL} to always pull all records in one batch. This will override the config value set on driver config.
+ * @param {string} database - The database this session will operate on.
+ * @param {string} impersonatedUser - The username which the user wants to impersonate for the duration of the session.
+ * @param {BookmarkManager} [bookmarkManager] = The bookmark manager
+ */
+class SessionConfig {
```
Comment (fbiville) [inline packages/core/src/driver.ts:106]: isn't going from a raw object to a class a breaking change? (which is OK for 5.0 anyway, just wondering)

File: packages/neo4j-driver/test/index.test.js [hunk 39]
```diff
@@ -139,7 +139,7 @@ describe('#unit index', () => {
       }
     })
 
-    describe('await rxRession.beginTransaction().toPromise() return', () => {
+    xdescribe('await rxRession.beginTransaction().toPromise() return', () => {
```
Comment (fbiville) [inline packages/neo4j-driver/test/index.test.js]: is that related to the other PR with rx changes?

**Review by fbiville — APPROVED** [review]

File: packages/neo4j-driver/test/index.test.js (hunk 39 above)
Comment (bigmontz) [reply packages/neo4j-driver/test/index.test.js]: ops, it is a leftover. I was skipping the test here for checking for another issues.

File: packages/core/src/driver.ts (hunk 38 above)
Comment (bigmontz) [reply packages/core/src/driver.ts:106]: Nope, typescript has loose type. This class will behave an interface actually, I need to define as class to make it possible add the experimental annotations in place. 

See usage example: https://github.com/neo4j/neo4j-javascript-driver/pull/974/files#diff-0ec3f1aeef14ced21b829fd32b84fb3a078e728e5fe474d34b33912dee2b7a1bR124

File: packages/core/src/internal/connection-holder.ts [hunk 40]
```diff
@@ -161,12 +163,16 @@ class ConnectionHolder implements ConnectionHolderInterface {
   }
 
   private async _createConnectionPromise (connectionProvider: ConnectionProvider): Promise<Connection | null> {
+    this._connectionError = undefined
```
Comment (fbiville) [inline packages/core/src/internal/connection-holder.ts]: you do not need to reset it here, do you?

File: packages/core/src/transaction.ts [hunk 41]
```diff
@@ -154,6 +159,10 @@ class Transaction {
         }
         this._onError(error).catch(() => {})
       })
+      // It should make the transaction active anyway
+      // futher errors will be treated by the exiting
```
Comment (fbiville) [inline packages/core/src/transaction.ts]: cannot edit for some reason:
```
// further errors will be treated by the existing
```

File: packages/core/src/transaction.ts [hunk 42]
```diff
@@ -347,31 +364,34 @@ const _states = {
         reactive,
         fetchSize,
         highRecordWatermark,
-        lowRecordWatermark
+        lowRecordWatermark,
+        preparationJob
       }: StateTransitionParams
     ): any => {
       // RUN in explicit transaction can't contain bookmarks and transaction configuration
       // No need to include mode and database name as it shall be inclued in begin
```
Comment (fbiville) [inline packages/core/src/transaction.ts]: old typo but: should be `included` instead of `inclued`

File: packages/core/test/transaction.test.ts [hunk 43]
```diff
@@ -343,11 +345,36 @@ function testTx<T extends Transaction> (transactionName: string, newTransaction:
...
 
+        tx._begin(async () => Bookmarks.empty(), TxConfig.empty())
+
         const result = tx.run('RETURN 1')
 
         // @ts-expect-error
         expect(result._watermarks).toEqual({ high: 700, low: 300 })
       })
+
+      it('should wait begin message be send', async () => {
```
Comment (fbiville) [inline packages/core/test/transaction.test.ts]: s/send/sent/

File: packages/core/test/transaction.test.ts [hunk 44]
```diff
@@ -376,6 +405,29 @@ function testTx<T extends Transaction> (transactionName: string, newTransaction:
             expect(error).toEqual(expectedError)
           }
         })
+
+        it('should wait begin message be send', async () => {
```
Comment (fbiville) [inline packages/core/test/transaction.test.ts]: s/send/sent/

**Review by fbiville — APPROVED** [review]

File: packages/core/src/internal/connection-holder.ts (hunk 40 above)
Comment (bigmontz) [reply packages/core/src/internal/connection-holder.ts]: I removed the changes in the connection-holder, to avoid effects in other areas.

**Review by robsdedude — APPROVED** [review]
:stopwatch:

File: packages/core/test/transaction.test.ts (hunk 43 above)
Comment (robsdedude) [inline packages/core/test/transaction.test.ts]:
```suggestion
      it('should wait until begin message was sent', async () => {
```

Comment (robsdedude) [reply packages/core/test/transaction.test.ts]: (Once more below)

**Comment by bigmontz** [comment]
Depends on: https://github.com/neo4j-drivers/testkit/pull/513
